### PR TITLE
Implement spread for objects

### LINKED ISSUE #214: Implement Spread operator for Objects
The rest operator has been implemented to arrays. `[a,b,...c]`
But it is not implemented for objects.

## test case

```js
const b = {
  a: 1,
  b: 2,
  c: 3
};

const a = {
  a: "a",
  b: "b",
  ...b
};

console.log(a);

```
The array implementation exists here inside of exec.rs:
https://github.com/jasonwilliams/boa/blob/master/src/lib/exec.rs#L218-L223

Tests for this exist here too:
https://github.com/jasonwilliams/boa/blob/master/src/lib/exec.rs#L850-L862

https://github.com/jasonwilliams/boa/blob/master/src/lib/exec.rs#L202-L213 will need to do something similar.

There will need to be changes to the parser so spread can be parsed within an object. The above code will currently error in the parser as spread syntax isn't expecting within an object.
Parsing objects happens here:
https://github.com/jasonwilliams/boa/blob/master/src/lib/syntax/parser.rs#L522-L574

### PR DESCRIPTION
**Pull request by FrancisMurillo**
This Pull Request fixes/closes #214.

It changes the following:

- Add spread operator for object initializer

### CODE REVIEW

**Comment by Razican** [comment]
### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,897 | 78,897 | 0 |
  | Passed | 26,974 | 26,990 | **+16** |
  | Ignored | 15,616 | 15,616 | 0 |
  | Failed | 36,307 | 36,291 | **-16** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 34.19% | 34.21% | **+0.02%** |
  
  <details><summary><b>Fixed tests:</b></summary>
  
  ```
  test/language/expressions/new/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/new/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/new/spread-obj-override-immutable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-override-immutable.js (previously Failed)
  test/language/expressions/array/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/array/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/call/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/call/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/call/spread-obj-override-immutable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-override-immutable.js (previously Failed)
  ```
  </details>

File: boa/src/syntax/ast/node/object/mod.rs [hunk 1]
```diff
@@ -146,7 +146,34 @@ impl Executable for Object {
...
+                                obj.set_property(
+                                    key.clone(),
+                                    PropertyDescriptor::builder()
+                                        .value(property)
+                                        .writable(true)
+                                        .enumerable(true)
+                                        .configurable(true),
+                                );
+                            }
+                        }
```
Comment (raskad) [inline boa/src/syntax/ast/node/object/mod.rs:173]: This is the `CopyDataProperties` function right? We have this implemented on `JsObject`.
```suggestion
                    let o = obj.as_object()?;
                    o.copy_data_properties(val, vec![], context)?;
```

**Review by raskad — COMMENTED** [review]
One suggestion otherwise looks good to me!

File: boa/src/syntax/ast/node/object/mod.rs (hunk 1 above)
Comment (FrancisMurillo) [reply boa/src/syntax/ast/node/object/mod.rs]: Nice, this makes it more succinct.

File: boa/src/exec/tests.rs [hunk 2]
```diff
@@ -118,6 +118,26 @@ fn object_field_set() {
...
+    let one = forward(&mut context, "a.x");
+    assert_eq!(one, String::from("-1"));
+
+    let two = forward(&mut context, "a.y");
+    assert_eq!(two, String::from("2"));
+
+    let three = forward(&mut context, "a.z");
+    assert_eq!(three, String::from("-3"));
+}
+
```
Comment (jedel1043) [inline boa/src/exec/tests.rs:140]: We have a new API for testing simple repetitive cases like this one. You can check it out on:

https://github.com/boa-dev/boa/blob/ae73c261a39f2b3a3d56e8b47d4ca289f56df87c/boa/src/lib.rs#L173-L203

Could you refactor this test to use this API? The other tests are ok.

**Review by raskad — APPROVED** [review]
Very nice work!
These are the latest test results:

### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 80,685 | 80,685 | 0 |
  | Passed | 32,670 | 32,754 | **+84** |
  | Ignored | 15,818 | 15,818 | 0 |
  | Failed | 32,197 | 32,113 | **-84** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 40.49% | 40.59% | **+0.10%** |
  
  <details><summary><b>Fixed tests (84):</b></summary>
  
  ```
  test/language/expressions/array/spread-obj-spread-order.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-spread-order.js (previously Failed)
  test/language/expressions/array/spread-obj-getter-init.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-getter-init.js (previously Failed)
  test/language/expressions/array/spread-obj-override-immutable.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-override-immutable.js (previously Failed)
  test/language/expressions/array/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/array/spread-obj-mult-spread.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-mult-spread.js (previously Failed)
  test/language/expressions/array/spread-obj-mult-spread-getter.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-mult-spread-getter.js (previously Failed)
  test/language/expressions/array/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/array/spread-obj-overrides-prev-properties.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-overrides-prev-properties.js (previously Failed)
  test/language/expressions/array/spread-mult-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-mult-obj-ident.js (previously Failed)
  test/language/expressions/array/spread-obj-getter-descriptor.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-getter-descriptor.js (previously Failed)
  test/language/expressions/array/spread-obj-with-overrides.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-with-overrides.js (previously Failed)
  test/language/expressions/array/spread-obj-manipulate-outter-obj-in-getter.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-manipulate-outter-obj-in-getter.js (previously Failed)
  test/language/expressions/array/spread-sngl-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-sngl-obj-ident.js (previously Failed)
  test/language/expressions/array/spread-obj-symbol-property.js [strict mode] (previously Failed)
  test/language/expressions/array/spread-obj-symbol-property.js (previously Failed)
  test/language/expressions/call/spread-obj-spread-order.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-spread-order.js (previously Failed)
  test/language/expressions/call/spread-obj-getter-init.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-getter-init.js (previously Failed)
  test/language/expressions/call/spread-obj-override-immutable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-override-immutable.js (previously Failed)
  test/language/expressions/call/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/call/spread-obj-mult-spread.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-mult-spread.js (previously Failed)
  test/language/expressions/call/spread-obj-mult-spread-getter.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-mult-spread-getter.js (previously Failed)
  test/language/expressions/call/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/call/spread-obj-overrides-prev-properties.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-overrides-prev-properties.js (previously Failed)
  test/language/expressions/call/spread-mult-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-mult-obj-ident.js (previously Failed)
  test/language/expressions/call/spread-obj-getter-descriptor.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-getter-descriptor.js (previously Failed)
  test/language/expressions/call/spread-obj-with-overrides.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-with-overrides.js (previously Failed)
  test/language/expressions/call/spread-obj-manipulate-outter-obj-in-getter.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-manipulate-outter-obj-in-getter.js (previously Failed)
  test/language/expressions/call/spread-sngl-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-sngl-obj-ident.js (previously Failed)
  test/language/expressions/call/spread-obj-symbol-property.js [strict mode] (previously Failed)
  test/language/expressions/call/spread-obj-symbol-property.js (previously Failed)
  test/language/expressions/new/spread-obj-spread-order.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-spread-order.js (previously Failed)
  test/language/expressions/new/spread-obj-getter-init.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-getter-init.js (previously Failed)
  test/language/expressions/new/spread-obj-override-immutable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-override-immutable.js (previously Failed)
  test/language/expressions/new/spread-err-sngl-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-err-sngl-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/new/spread-obj-mult-spread.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-mult-spread.js (previously Failed)
  test/language/expressions/new/spread-obj-mult-spread-getter.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-mult-spread-getter.js (previously Failed)
  test/language/expressions/new/spread-err-mult-err-obj-unresolvable.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-err-mult-err-obj-unresolvable.js (previously Failed)
  test/language/expressions/new/spread-obj-overrides-prev-properties.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-overrides-prev-properties.js (previously Failed)
  test/language/expressions/new/spread-mult-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-mult-obj-ident.js (previously Failed)
  test/language/expressions/new/spread-obj-getter-descriptor.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-getter-descriptor.js (previously Failed)
  test/language/expressions/new/spread-obj-with-overrides.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-with-overrides.js (previously Failed)
  test/language/expressions/new/spread-obj-manipulate-outter-obj-in-getter.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-manipulate-outter-obj-in-getter.js (previously Failed)
  test/language/expressions/new/spread-sngl-obj-ident.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-sngl-obj-ident.js (previously Failed)
  test/language/expressions/new/spread-obj-symbol-property.js [strict mode] (previously Failed)
  test/language/expressions/new/spread-obj-symbol-property.js (previously Failed)
  ```
  </details>

**Review by jedel1043 — CHANGES_REQUESTED** [review]
The implementation looks good! Just a cleanup of a test to use the new API and this should be ready.

**Review by jedel1043 — APPROVED** [review]
Good work!

**Pull request merged**